go/vt/sqlparser/ast.go

```diff
@@ -2707,6 +2707,34 @@ type (
 	}
 
 	CountStar struct {
+		Name string
+		_    bool
+		// TL;DR; This makes sure that reference equality checks works as expected
+		//
+		// You're correct that this might seem a bit strange at first glance.
+		// It's a quirk of Go's handling of empty structs. In Go, two instances of an empty struct are considered
+		// identical, which can be problematic when using these as keys in maps.
+		// They would be treated as the same key and potentially lead to incorrect map behavior.
+		//
+		// Here's a brief example:
+		//
+		// ```golang
+		// func TestWeirdGo(t *testing.T) {
+		// 	type CountStar struct{}
+		//
+		// 	cs1 := &CountStar{}
+		// 	cs2 := &CountStar{}
+		//  if cs1 == cs2 {
+		// 	  panic("what the what!?")
+		//  }
+		// }
+		// ```
+		//
+		// In the above code, cs1 and cs2, despite being distinct variables, would be treated as the same object.
+		//
+		// The solution we employed was to add a dummy field `_ bool` to the otherwise empty struct `CountStar`.
+		// This ensures that each instance of `CountStar` is treated as a separate object,
+		// even in the context of out semantic state which uses these objects as map keys.
 	}
```

Comment:
Noting that I didn't include [the `OverClause` field from the upstream change](https://github.com/vitessio/vitess/blob/d3c0d255c405b6fd2468f0053661a361a81e6ebd/go/vt/sqlparser/ast.go#L2927) - v15 doesn't support the `OverClause` support

Reply:
Context on `OverClause`: https://github.com/vitessio/vitess/pull/13444